uktrade/great-cms
# GP2-162: Module page lesson-completion indicator

This changeset adds support for showing, on the module page (`CuratedListPage`) when its configured lessons have been completed.

Significant changes:
* The helper code to get read progress has been renamed to show it reflects lesson-completion status
* The helper now ONLY shows details of lessons/`DetailPage`s that are configured in the `CuratedListPage`'s `topics` field, rather than all `DetailPage`s which are direct descendents of the `CuratedListPage` -- this fixes the total-lessons count by ignoring unconfigured child `DetailPage`s
* The helper now also returns an extra dict, showing topics with completed lessons:

Original data structure

```
{
    'lessons_in_progress': True,
    'module_pages': [
        {
            'page': <Page: Identify opportunities and research the market>,
            'read_count': 4,
            'total_pages': 9  # not all were associated with topics, only 7
        },
        {
            'page': <Page: Prepare to sell in a new country>,
            'read_count': 2,
            'total_pages': 5  # again, one too many
        },
        {
            'page': <Page: Selling across borders: product and services regulations, licensing and logistics>,
            'read_count': 0,
            'total_pages': 1
        }
    ]
}
```

New data structure:
```
{
    'lessons_in_progress': True,
    'module_pages': [
        {
            'total_pages': 7,  # Now correct
            'completion_count': 4,
            'page': <Page: Identify opportunities and research the market>,
            'completed_lesson_pages': {
                'b7eca1bf-8b43-4737-91e4-913dfeb2c5d8': {10, 26},
                '044e1343-f2ce-4089-8ce9-17093b9d36b8': {18, 20}}
            )
        },
        {
            'total_pages': 4,  # Now correct
            'completion_count': 2,
            'page': <Page: Prepare to sell in a new country>,
            'completed_lesson_pages': {
                    '786e9140-6ba8-4f1a-970b-0d556013e64d': {14, 22}
                }
            )
        }
    ],
}
```
 
<img width="1389" alt="Screenshot 2020-10-21 at 21 41 11" src="https://user-images.githubusercontent.com/101457/96789743-a157e100-13ed-11eb-9cd7-e0b5a973ecf0.png">
<img width="1302" alt="Screenshot 2020-10-21 at 21 30 58" src="https://user-images.githubusercontent.com/101457/96789750-a3ba3b00-13ed-11eb-8595-9362db52657c.png">
<img width="1284" alt="Screenshot 2020-10-21 at 21 27 29" src="https://user-images.githubusercontent.com/101457/96789751-a4eb6800-13ed-11eb-9a9f-57924476e7ae.png">

 - [X] Ticket exists in Jira  https://uktrade.atlassian.net/browse/GP2-162 

 - [X] How to test locally, including how to set up appropriate data
 
 
### To test
* Add a couple of placeholder records to the Topics within a module (optional, but recommended)
* Load up the dashboard page and confirm it now shows the correct numbers for total lessons in each module. (Note that modules without any lessons configured via topics will not show at all.)
* Load up a lesson module page in a new tab
* Open a lesson from that module in a new tab.
* As you work through lessons, tick each as completed, one at a time
* Before you move on to the next lesson, refresh the dashboard and confirm the completion counter increases appropriately. Also refresh the module page and confirm the lesson is marked as completed.
